gson/src/main/java/com/google/gson/internal/bind/TreeTypeAdapter.java

```diff
@@ -92,7 +92,15 @@ public T read(JsonReader in) throws IOException {
     if (nullSafe && value.isJsonNull()) {
       return null;
     }
-    return deserializer.deserialize(value, typeToken.getType(), context);
+
+    try {
+      return deserializer.deserialize(value, typeToken.getType(), context);
+    } catch (Exception e) {
```

Comment:
My main concern here is that `JsonDeserializer` is an interface that users can and do implement themselves. That means we don't know what sort of unchecked exception might be thrown here, and by changing it to `JsonParseException` we might change the functioning of existing code. Perhaps if we only catch `JsonParseException`?

I'm going to run this change through all of Google's tests to see what happens.

Reply:
> Perhaps if we only catch `JsonParseException`?

Sounds reasonable, that will then also cover the subclasses `JsonIOException` and `JsonSyntaxException`.

And thanks for testing this!

However, my concern is still that users might only use `JsonDeserializer` and then could end up with for example 10 chained exceptions when deserialization fails for a nested value.
So I really wouldn't mind if this change is not integrated, if the inconvenience of incomplete JSON path information does not justify risking other issues.

Reply:
So the test results are in, and it's not great. Over 30 failures spread across a number of different projects. Quite a lot of tests looking for the exact exception message you get now, which is arguably overspecific. But some others expecting `JsonSyntaxException` where now they get `JsonParseException`. In some cases it is just `assertThrows` or the like, but in others people are catching only `JsonSyntaxException` in production code, and that no longer works if we rethrow as `JsonParseException`. While we can imagine various workarounds, I think this change is probably more trouble than it's worth?